### PR TITLE
Feature/84 es6 server

### LINKED ISSUE #84: Allow for ES6 development against td.server
**Describe what problem your feature request solves**
`td.server` should allow for ES6 development

**Describe the solution you'd like**
Using babel or a similar technology, the back-end code should be able to handle es6 syntax, such as `import './foobar.js'`

**Additional context**
Rather than doing a large refactor all at once, I think it would make sense to create the infrastructure to support ES6 development first, and then chip away at refactoring existing code.

### LINKED ISSUE #91: Use nyc for istanbul
**Describe what problem your feature request solves**
Use [nyc](https://github.com/istanbuljs/nyc) for istanbul coverage

**Additional context**
When installing dependencies via npm, there is a deprecation warning for istanbul that instructs the user to use nyc or similar as istanbul is no longer maintained.

### PR DESCRIPTION
**Pull request by lreading**
**Summary**
Adding ability to write server code using ES6, as well as tests.

**Description for the changelog**
- Added babel and transpilation script to backend: Closes #84 
- Updated Dockerfile to use transpiled output
- Removed istanbul in favor of nyc for server: Closes #91 
- Added mocha, sinon and chai for unit testing, removed jasmine/mockery
- Refactored a large amount of tests to work with mocha/sinon/chai
- Refactored some server code to improve testability
- Refactored dotenv config file to use ES6 syntax

**Other info**
I had run into some problems using babel with jasmine so it made sense to pull in mocha/sinon/chai, as is listed on the roadmap. There are still some tests that can be refactored to use es6 syntax.  
I did, intentionally, shy away from using supertest during the refactor.  My experience with supertest is that the tests wind up becoming integration tests moreso than unit tests.  I think that having a separate integration test suite would be a nice addition, especially when we begin working on the API.

I'm not sure how this will play with codecov quite yet, as it's not writing anything to the `coverage/Server` directory like istanbul used to.  I've never used codecov before; any help/guidance would be appreciated!

I apologize for how... aggressive... this PR is.  I started out with the intention of keeping it as small as possible and clearly failed. :sweat_smile:

### CODE REVIEW

**Comment by jgadsden** [comment]
No problem at all with the extent of this PR @lreading - very good to see.  I will try it out on Linux, MacOS and even dig out my Windows machine to double confirm that all is working as we expect. There have been a lot of changes recently and it would be good to do a thorough check with this PR.

@andk123 or @mike-goodwin you are better placed to test on Windows, are you able to check this PR out yourselves?

I suppose I am saying thatit might take me a while to review the PR @lreading , so bear with me :-)

**Comment by jgadsden** [comment]
Not really associated with this PR, but something that came up during testing, is that if `npm install` is not run first then there is an error when building the docker image using  `docker build -t owasp-threat-dragon:dev .` : 

```
npm ERR! cipm can only install packages with an existing package-lock.json or npm-shrinkwrap.json with lockfileVersion >= 1. Run an install with npm@5 or later to generate it, then try again.
```
Nobody should try and build the docker image without first running `npm install` ... but having said that someone will. It would be good to put a note or something to this effect in the README.md and docs, just so that we do not get issues raised on this.

@lreading shall I raise an issue for this? Or do you want to add it in this PR? or raise an issue yourself? I am OK with any of these :-)

**Comment by lreading** [comment]
> Not really associated with this PR, but something that came up during testing, is that if `npm install` is not run first then there is an error when building the docker image using `docker build -t owasp-threat-dragon:dev .` :
> 
> ```
> npm ERR! cipm can only install packages with an existing package-lock.json or npm-shrinkwrap.json with lockfileVersion >= 1. Run an install with npm@5 or later to generate it, then try again.
> ```
> 
> Nobody should try and build the docker image without first running `npm install` ... but having said that someone will. It would be good to put a note or something to this effect in the README.md and docs, just so that we do not get issues raised on this.
> 
> @lreading shall I raise an issue for this? Or do you want to add it in this PR? or raise an issue yourself? I am OK with any of these :-)

That's an interesting find!  [NPM recommends committing package-lock.json to source control](https://docs.npmjs.com/cli/v7/configuring-npm/package-lock-json).

I would argue that the docker build process should be able to be run without any dependencies external to docker (eg running npm install first).  The simple solution in my opinion is adding `package-lock.json` to source control. 

Thoughts?

**Comment by jgadsden** [comment]
> I would argue that the docker build process should be able to be run without any dependencies external to docker (eg running npm install first). The simple solution in my opinion is adding `package-lock.json` to source control.
> 
> Thoughts?

True ... and it was useful to be prompted to run `npm install` in both directories ... so the error was quite useful in an unintended way. 

Having said that I agree that you are correct, and that the `package-lock.json` should be added to the source repo.

**Comment by lreading** [comment]
I'll update it to include the lock in both td.site and td.server.
> ... and it was useful to be prompted to run npm install in both directories ... so the error was quite useful in an unintended way.
You should be able to run it once at the root.  There's still a package.json there, and when running `install` from the root there's a `postinstall` script that will run install in both site/server.

I'll review the readme as well to make sure that's clear, I think that may have been overlooked. :)

**Review by jgadsden — APPROVED** [review]
All test out perfectly @lreading . Just one question I am having with the Dockerfile, but I will put that in a separate comment.

**Comment by jgadsden** [comment]
Having problems running the docker container after having rebuilt the docker image from this branch. 

In PR94 branch: 
1. set up `.env` file similar to `example.env`
1. `docker build -t owasp-threat-dragon:dev . `
2.`docker run -it -p 3000:3000 -v $(pwd)/.env:/app/td.server/.env owasp-threat-dragon:dev`
3. throws error at  me : 
```
pr94$ docker run -it -p 3000:3000 -v $(pwd)/.env:/app/td.server/.env owasp-threat-dragon:dev
/app/td.server/dist/helpers/express.helper.js:1
import express from 'express';
^^^^^^
SyntaxError: Cannot use import statement outside a module
    at wrapSafe (internal/modules/cjs/loader.js:1152:16)
    at Module._compile (internal/modules/cjs/loader.js:1200:27)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1257:10)
    at Module.load (internal/modules/cjs/loader.js:1085:32)
    at Function.Module._load (internal/modules/cjs/loader.js:950:14)
    at Module.require (internal/modules/cjs/loader.js:1125:19)
    at require (internal/modules/cjs/helpers.js:75:18)
    at Object.<anonymous> (/app/td.server/dist/app.js:7:21)
    at Module._compile (internal/modules/cjs/loader.js:1236:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1257:10)
```

In main branch : 
1. set up `.env` file similar to `example.env`
1. `docker build -t owasp-threat-dragon:dev . `
2.`docker run -it -p 3000:3000 -v $(pwd)/.env:/app/td.server/.env owasp-threat-dragon:dev`
3. throws a different error at me using the same `.env` file as used in PR94:
```
pr94$ docker run -it -p 3000:3000 -v $(pwd)/.env:/app/td.server/.env owasp-threat-dragon:dev
{"name":"threatdragon","excludes":["req-headers","res-headers","res","req","short-body","body","response-hrtime","incoming","user-agent","response-time","http-version"],"hostname":"3a6b86c1252f","pid":1,"level":50,"security":true,"msg":"secure session cookie flag was false - should only happen in dev environments","time":"2021-04-30T16:26:01.871Z","v":0}
{"name":"threatdragon","hostname":"3a6b86c1252f","pid":1,"level":50,"msg":"owasp threat dragon failed to start up","time":"2021-04-30T16:26:01.875Z","v":0}
{"name":"threatdragon","hostname":"3a6b86c1252f","pid":1,"level":50,"msg":"Credentials must be provided when creating a service client.","time":"2021-04-30T16:26:01.876Z","v":0}
Express server listening on port 3000
```
Not sure why the creds are not found, these are the same creds used successfully when not running in docker container

Do you want this to be a separate issue @lreading ?

**Comment by lreading** [comment]
I don't like the idea of merging anything to the main branch that might break something, so I'd prefer to handle it here. :) 

I'll take a look at both issues and try to figure out where I screwed up! :stuck_out_tongue: 

Thanks for checking so thoroughly! It's nice to have people willing to take the time to look to make sure things are working as expected.  The "works on my machine" approach doesn't always cut it. :wink:

File: package-lock.json [hunk 1]
```diff
@@ -0,0 +1,1366 @@
+{
```
Comment (jgadsden) [inline package-lock.json:1]: having this package-lock.json should also sort out dependabot security alert : https://github.com/OWASP/threat-dragon/security/dependabot/td.site/package.json/pug/open

**Comment by jgadsden** [comment]
The docker build and run is giving me problems on MacOS. I cleared out all old images and containers from docker and rebuilt from scratch, but sill get this error. @lreading I am not sure if you have MacOS test machine, so I can look at this if you like? Seems to be breaking on port value.

```
84-es6-server$ docker run -it -p 3000:3000 -v $(pwd)/.env:/app/.env owasp-threat-dragon:dev
{"name":"threatdragon","excludes":["req-headers","res-headers","res","req","short-body","body","response-hrtime","incoming","user-agent","response-time","http-version"],"hostname":"c137f07aa1ba","pid":1,"level":50,"security":true,"msg":"secure session cookie flag was false - should only happen in dev environments","time":"2021-05-02T07:23:26.103Z","v":0}
{"name":"threatdragon","hostname":"c137f07aa1ba","pid":1,"level":50,"msg":"owasp threat dragon failed to start up","time":"2021-05-02T07:23:26.107Z","v":0}
{"name":"threatdragon","hostname":"c137f07aa1ba","pid":1,"level":50,"msg":"Credentials must be provided when creating a service client.","time":"2021-05-02T07:23:26.107Z","v":0}
/app/td.server/index.js:5
app.set('port', process.env.PORT || 3000);
    ^
TypeError: Cannot read property 'set' of null
    at Object.<anonymous> (/app/td.server/index.js:5:5)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:72:12)
    at internal/main/run_main_module.js:17:47
```

**Comment by lreading** [comment]
@jgadsden I don't have a mac to test it on at the moment... I have one that needs rebuilding and hopefully I'll get that and at least a windows VM set up this week so I can cross-platform test.

That said, I'm going to try to reproduce on a machine that has no existing images or anything.  If it's happening within a docker context, in THEORY, it shouldn't be platform dependent.

**Comment by lreading** [comment]
This is another issue with the `.env` file.  `app.js` now returns the express app, and if it fails to create it, it returns null.  This is why the error is that it can't read property set of **null** as opposed to undefined.  When I run it without a `.env` file (and without exporting any variables) I get the same error.

When I run it with the correctly configured `.env` file, it works as expected.  I think this ties back to the conversation about configuration and better handling of secrets.  It'd be nice to have a more meaningful error message for this scenario.

Is your `.env` file configured correctly, and are you running it from the same directory where the `.env` file lives?

**Comment by jgadsden** [comment]
aha! Thanks @lreading I found was exporting the env variables, in effect  my .env looked like : 
```
export GITHUB_CLIENT_ID=01234567890123456789
export GITHUB_CLIENT_SECRET=0123456789abcdef0123456789abcdef0123456
export NODE_ENV=development
export SESSION_STORE=local
export SESSION_SIGNING_KEY=11223344556677889900aabbccddeeff
export SESSION_ENCRYPTION_KEYS='[{"isPrimary": true, "id": 0, "value": "11223344556677889900aabbccddeeff"}]'
export GITHUB_SCOPE=public_repo
```

This is the same file as I use when running from the command line, `npm start`, which works OK from the command line but not when running within the docker container ... odd as I expected dotenv to be OK with `export`. Anyhow I removed all the `export` from the `.env` and it is now working as you have it.

The reason why I have the exports is because in `setup-env.md` we specify exporting the env vars ... this is probably historical as I am not sure if we _really_ need to export them.

**Review by jgadsden — APPROVED** [review]

**Comment by jgadsden** [comment]
Also checked that it is working on my Linux (it is) and also on Windows - so all good from my side if you want to go ahead and merge @lreading

**Pull request merged**